test/integration/outputcheck/outputcheck.pony

```diff
@@ -0,0 +1,127 @@
+  success to ```make```. If is not what is expected, or the child process ends
+  with a non-zero error code, a non-zero error code is returned by parent.
+
+  This leads to code like this in the Main actor:
+  ```
+    if env.args.size() > 1
+    then
+      // we're the child/test process
+      // ADD CODE HERE TO PERFORM THE TEST, 
+      // WITH OUTPUT SHOWING TEST PASS / FAIL
+    else
+      // we're the parent driver/assert process
+      let expected = recover val
+        [ SET OF STRINGS THAT ARE EXPECTED TO BE OUTPUT ]
+      end
+      OutputCheck(expected).run(env)
+    end
+  ```
+
+  Currently verification of the actual and expected output is based on them
+  being the same *set* of line strings. This could obviously be extended to
+  consider ordering.
+
+  Standard error output from the child process is echoed to the parent process's
+  stdout.
+  """
+  let _expected : Array[String] val
+
+  new create(expectedResults: Array[String] val) =>
+    _expected = expectedResults
+
+  be run(env: Env) =>
+    let client = _ProcessClient(env, _expected)
+    let processFilename: String = try env.args(0) else "" end
+    // define the binary to run
+    try
+      let path = FilePath(env.root as AmbientAuth, 
+                          processFilename)
+
+      // define the arguments; first arg is always the binary name
+      let args: Array[String] iso = recover [processFilename; "dummy to avoid reforking"] end
+
+      // create a ProcessMonitor and spawn the child process
+      let auth = env.root as AmbientAuth
+      let pm: ProcessMonitor = ProcessMonitor(auth, 
+        consume client, path,
+        consume args, recover val Array[String]() end)
+    else
+      env.out.print("Could not create FilePath!")
+    end
+
+
+class _ProcessClient is ProcessNotify
+  let _env: Env
+  let _expected: Array[String] val
+  let _actual: Reader = Reader
+
+  new iso create(env: Env, expected: Array[String] val) =>
+    _env = env
+    _expected = expected
+
+  fun ref stdout(process: ProcessMonitor ref, data: Array[U8] iso) =>
+    _actual.append(consume data)
+
+  fun ref stderr(process: ProcessMonitor ref, data: Array[U8] iso) =>
+    let err = String.from_array(consume data)
+    _env.out.print("STDERR: " + err)
+
+  fun ref failed(process: ProcessMonitor ref, err: ProcessError) =>
+    match err
+    | ExecveError => _env.out.print("ProcessError: ExecveError")
+    | PipeError => _env.out.print("ProcessError: PipeError")
+    | ForkError => _env.out.print("ProcessError: ForkError")
+    | WaitpidError => _env.out.print("ProcessError: WaitpidError")
+    | WriteError => _env.out.print("ProcessError: WriteError")
+    | KillError => _env.out.print("ProcessError: KillError")
+    | CapError => _env.out.print("ProcessError: CapError")
+    | Unsupported => _env.out.print("ProcessError: Unsupported")
+    else _env.out.print("Unknown ProcessError!")
+    end
+    _exit(1)
+
+  fun _exit(code: I32) =>
+    @_exit[None](I32(code))
```

Comment:
Early program termination via `exit` or similar is generally a bad idea in Pony because it doesn't allow the runtime to properly cleanup resources before terminating.

The approach used in Pony is to set the exit code with `pony_exitcode`. This doesn't cause the runtime to immediately terminate, you still have to reach quiescence by not sending any new messages and closing any existing I/O connection.